src/ir/auth_logic/ast_construction.cc

```diff
@@ -128,6 +128,40 @@ static Fact ConstructFact(std::forward_list<Principal> delegation_chain,
                        *CHECK_NOTNULL(can_say_fact_context.fact()));
 }
 
+static ArgumentType ConstructArgumentType(
+    AuthLogicParser::AuthLogicTypeContext& auth_logic_type_context) {
+  if (auto* number_type_context =
+          dynamic_cast<AuthLogicParser::NumberTypeContext*>(
+              &auth_logic_type_context)) {
+    return ArgumentType(ArgumentType::Kind::kNumber);
+  } else if (auto* principal_type_context =
+                 dynamic_cast<AuthLogicParser::PrincipalTypeContext*>(
+                     &auth_logic_type_context)) {
+    return ArgumentType(ArgumentType::Kind::kPrincipal);
+  }
+  auto& custom_type_context =
+      *CHECK_NOTNULL(dynamic_cast<AuthLogicParser::CustomTypeContext*>(
+          &auth_logic_type_context));
+  return ArgumentType(ArgumentType::Kind::kCustom,
+                      custom_type_context.ID()->getText());
+}
+
+static RelationDeclaration ConstructRelationDeclaration(
+    AuthLogicParser::RelationDeclarationContext& relation_declaration_context) {
+  std::vector<std::pair<std::string, ArgumentType>> arguments;
+  for (uint64_t i = 0; i < relation_declaration_context.authLogicType().size();
+       ++i) {
+    arguments.push_back(std::pair(
+        relation_declaration_context.ID(i + 1)->getText(),
+        ConstructArgumentType(
+            *CHECK_NOTNULL(relation_declaration_context.authLogicType(i)))));
+  }
```

Comment:
This is pretty subtle and looks like it could use a comment explaining it. Why are we looking up `ID(i + 1)`? Is this a 1-indexed collection? It sets off my "maybe there's a bug" alarm bells to see `i+1` and `i` used to index into stuff in `relation_declaration_context` in the same loop, but it could be fine depending.

Reply:
Comments provided in the file explaining the need to `i` and `i+1`.